File: github_code/avl-tree/java/12.java

package avltree;

public class AvlTree {
    
    public long data;
    public AvlTree left;
    public AvlTree right;
    int height = 0;
    
    int height(AvlTree p){
        if ( p == null){
            return -1;
        }
        else{
            return p.height;
        }
    }
    
    public AvlTree insert(AvlTree p, int x){
        if ( p == null ){
            p = new AvlTree();
            p.data = x;
        }
        else{
            if(x < p.data ){
                p.left = insert(p.left, x);
                if( height(p.left) - height(p.right) == 2)
                    if( x < p.left.data)
                        p = singlerotateright(p);
                else
                        p = doublerotateright(p);
            }
            else if ( x > p.data){
                p.right = insert(p.right, x);
                if( height(p.left) - height(p.right) == -2)
                    if (x > p.right.data)
                        p = singlerotateleft(p);
                else
                        p = doublerotateleft(p);
            }
            p.height = max(height(p.left), height(p.right)) +1; 
        }
        return p;
    }
         int max ( int a, int b){
        if( a > b){
            return a;
        }
        else {
            return b;
        }
    }
         
      AvlTree singlerotateleft(AvlTree p){
        AvlTree lc = p.left;
        lc.right = p;
        p.height = max(height(p.left), height(p.right)) +1;
        lc.height = max(height(lc.left), lc.height) +1;
        return lc;
    }
      
      AvlTree doublerotateright(AvlTree p){
        p.left = singlerotateleft(p.left);
        return singlerotateright(p);
    }
      
      AvlTree singlerotateright(AvlTree p){
        AvlTree rc = p.right;
        p.right = rc.left;
        rc.left = p;
        p.height = max(height(p.left), height(p.right))+1;
        rc.height = max (height(rc.right), rc.height) +1;
        return rc;
    }
      
      AvlTree doublerotateleft(AvlTree p){
        p.right = singlerotateright(p.right);
        return singlerotateleft(p);
    }
   

    public static void main(String[] args) {
        // TODO code application logic here
    }
    
}

File: github_code/avl-tree/cpp/12.c
//avl tree 
#include <iostream>
using namespace std;

struct node {
	int info,height;
	node *left,*right;
};

node *getnode (int data) {
	node *newnode = new node;
	newnode->info = data;
	newnode->height = 0;
	newnode->left = newnode->right = NULL;
	return newnode;
}

int height (node *curr) {
	if (curr == NULL)
		return 0;
	return curr->height; 
}

int getBalance (node *curr) {
	if (curr == NULL)
		return 0;
	return (height(curr->left) - height(curr->right));
}

node *leftleft (node *y) {
	node *x = y->left;
	node *subx = x->right;
	x->right = y;
	y->left = subx;
	
	y->height = 1 + max(height(y->left),height(y->right));
	x->height = 1 + max(height(x->left),height(x->right));
	
	return x;
}

node *rightright (node *y) {
	node *x = y->right;
	node *subx = x->left;
	x->left = y;
	y->right = subx;
	
	y->height = 1 + max(height(y->left),height(y->right));
	x->height = 1 + max(height(x->left),height(x->right));
	
	return x;
}

node *insert (node *root,int data) {
	if (root == NULL) 
		root = getnode(data);		
	else if (root->info > data)
		root->left = insert(root->left,data);
	else if (root->info < data)
		root->right = insert(root->right,data);
	else 
		return root;
		
	root->height = 1 + max(height(root->left),height(root->right));
	int balance = getBalance(root);
	
	if (balance == 2 && data < root->left->info) {//LL
		return leftleft(root);
	}
	else if (balance == 2 && data > root->left->info) {//LR
		root->left = rightright(root->left);
		return leftleft(root);
	}
	else if (balance == -2 && data > root->right->info) {//RR
		return rightright(root);
	}
	else if (balance == -2 && data < root->right->info) {//RL
		root->right = leftleft(root->right);
		return rightright(root);
	}
	
	return root;
}

void inorder (node *root) {
	if (root == NULL)
		return;
	inorder(root->left);
	cout<<root->info<<" ";
	inorder(root->right);
}

int main () {
	node *root = NULL;
	root = insert(root,25);
	root = insert(root,20);
	root = insert(root,35);
	root = insert(root,30);
	root = insert(root,24);
	inorder(root);
	cout<<"\n";
}
